### ISSUE TITLE
Web GUI very slow in v5.2.1

### ISSUE THREAD

**Issue by mcrapts** [pi-hole/docker-pi-hole#772]
### Versions
  Pi-hole version is v5.2.1 (Latest: v5.2.1)
  AdminLTE version is v5.2.1 (Latest: v5.2.1)
  FTL version is v5.3.2 (Latest: v5.3.2)

### Platform
- OS and version:  Ubuntu 20.04 LTS
- Platform:  Docker

### Expected behavior

With v5.1 every page of the web GUI responds <200ms. The expected behavior is for v5.2 to match the performance of v5.1.

### Actual behavior / bug

With v5.2 the average response time of the web GUI is 3 seconds. This number was obtained by using the Chrome developer tools. Page load time with v5.2 was <200ms. On my laptop with i5 6300u the response time is 903ms.

### Steps to reproduce

1. Launch Pihole v5.2 using the docker-compose.yml from https://hub.docker.com/r/pihole/pihole
2. Open Chrome
3. Open Dev Tools
4. Load http://localhost/admin/index.php 
5. Open Network Tab in Dev Tools
6. Look at the top row and see that the load time is is almost 1 second (or more on slower systems)

## Debug Token

- URL:  https://tricorder.pi-hole.net/dg98dkp77c

## Additional context
I tested the performance on Ubuntu server with Intel Celeron N3050 and laptop with Intel Skylake i5 6300u. Both were noticably slower than v5.1. The slowness is more apparent on slower CPU's, but an Intel i5 should not have any problems with the Pihole web GUI.

**Comment by mcrapts** [comment]
The latest version (v5.5.1) still has this problem.

**Comment by mcrapts** [comment]
This problem still exists with v5.6.

**Comment by tjengbudi** [comment]
i also got this probem,
used Pi-hole v5.2.4 Web Interface v5.4 FTL v5.7

this problem occurs when i update the pihole image. after clear and run the image. the web interface very slow.

but when it is run first time. it is ok. run smoothly. run first time means create docker with no mapping config before. create fresh  docker instant. but after any update. i download, clear image. and run it. the web interface get very slow response.

any tips to resolve this problem?

**Comment by RixzZ** [comment]
Same thing here. I changed from a RPI3 with Pi-Hole installed directly to a RPI4 with Pi-Hole on Docker and the WEBUI is too slow now...

**Comment by XenHat** [comment]
I also have noticed this issue.
Pihole versions: `Pi-hole v5.2.4 Web Interface v5.4 FTL v5.7`

Relevant YML data
```yml
services:
  pihole:
    image: pihole/pihole:latest
    network_mode: "host"
environment:
      TZ: 'America/New_York'
      DNS1: '127.0.0.1#5555'
    volumes:
      - './etc-pihole/:/etc/pihole/'
      - './etc-dnsmasq.d/:/etc/dnsmasq.d/'
      - './etc-dnsmasq/:/etc/dnsmasq/'
      - './extra/start.sh:/start.sh'
      - './etc/hosts:/etc/hosts'
    cap_add:
      - NET_ADMIN
    restart: unless-stopped
```

Misc. host info:
systemd-resolved masked, using unbound running on the hose
```bash
$ cat /sys/firmware/devicetree/base/model
Raspberry Pi 4 Model B
$ uname -a
Linux rpi4 5.11.2-1-ARCH #1 SMP Sun Feb 28 21:51:14 UTC 2021 aarch64 GNU/Linux
$ cat /etc/resolv.conf
# Generated by resolvconf
domain homelab.*********
nameserver 127.0.0.1
```

**Comment by dschaper** [comment]
> ```yaml
>       - './etc-dnsmasq.d/:/etc/dnsmasq.d/'
>       - './etc-dnsmasq/:/etc/dnsmasq/'
> ```

That's probably not correct. There is an `/etc/dnsmasq.d` directory and an `/etc/dnsmasq.conf` but not an `/etc/dnsmasq/` folder:

```plain
/etc $ ls -la dnsmasq*
-rw-r--r-- 1 root root    24 Mar  9 14:51 dnsmasq.conf
-rw-r--r-- 1 root root 25997 Mar  9 14:51 dnsmasq.conf.old

dnsmasq.d:
total 12
drwxr-xr-x  2 root root 4096 Apr  7 17:43 .
drwxr-xr-x 89 root root 4096 Mar 26 04:28 ..
-rw-r--r--  1 root root 1380 Mar 26 04:10 01-pihole.conf
-rw-r--r--  1 root root    0 Apr  7 17:43 05-pihole-custom-cname.conf
```



> ```yaml
>       - './extra/start.sh:/start.sh'
> ```

What is that script doing?

**Comment by XenHat** [comment]
> > ```yaml
> >       - './etc-dnsmasq.d/:/etc/dnsmasq.d/'
> >       - './etc-dnsmasq/:/etc/dnsmasq/'
> > ```
> 
> That's probably not correct. There is an `/etc/dnsmasq.d` directory and an `/etc/dnsmasq.conf` but not an `/etc/dnsmasq/` folder:

It is correct. This is a remnant of using the pihole instead of the LanCache-DNS container, which creates this directory and refers to it. I have since disabled it but I did not change the container config ;)
> 
> What is that script doing?
This is the container's start script, which I modified to disable the cache. 
The issue still exists on an unmodified container but I decided to be completely transparent and include my _running_ configuration.

**Comment by dschaper** [comment]
Is the web interface still "slow" with an absolute stock docker container and not added or customized configs?

Can you quantify in some way what "slow" means?

**Comment by XenHat** [comment]
> Is the web interface still "slow" with an absolute stock docker container and not added or customized configs?
A brand new clone of the repository (I seem to have had a very old version of it...) and a clean configuration and database, with my configuration imported with the teleporter does not exhibit **unbearable** slowdowns.
The FTL interface appears to load within 1-2 seconds which is a little slower than on the host directly (instant-0.5 seconds).

> Can you quantify in some way what "slow" means?

I cannot measure the old behaviour anymore as I made a mistake moving the databases and I have lost all the entries and requests, but it was on the order of **3-5 seconds** to load another page in the FTL interface. Which page **did not** seem to matter.

DNS requests themselves did not seem to slow down at all.

**Comment by mcrapts** [comment]
> Is the web interface still "slow" with an absolute stock docker container and not added or customized configs?
> 
> Can you quantify in some way what "slow" means?

Yes. Slow means that the "Query log" page loads at least 10x slower measured using Chrome dev tools. Compare any recent version with `pihole/pihole:v5.1.2` and I am sure you will measure this 10x difference as well.

**Comment by RixzZ** [comment]
I can add a small video to show it...

https://user-images.githubusercontent.com/1339272/114091771-5f4a2b00-98b9-11eb-8051-017efc6f7b8f.mp4

The slowness is the same accessing any page of the webui.

**Comment by XenHat** [comment]
> EDIT: I appeared to have read over "not added or customized", I will wipe the internal pihole config and measure again

Completely empty configuration seems to load the same as my imported one, FYIY.

**Comment by XenHat** [comment]
> I can add a small video to show it...
> 
>  Peek.2021-04-08.22-24.mp4 
> The slowness is the same accessing any page of the webui.

Yes. The loading speed is very much like what I was experiencing :+1:

**Comment by dschaper** [comment]
> Compare any recent version with `pihole/pihole:v5.1.2` and I am sure you will measure this 10x difference as well.

I don't.

**Comment by dschaper** [comment]
> I can add a small video to show it...

Show your configuration please.

**Comment by dschaper** [comment]
I guess I'm not understanding where there are huge slowdowns. Obviously running in a virtualized environment like docker is not going to get you bare metal performance but I'm not seeing 5 second page loads or 10x delays. 

But I'll leave this to the docker people that can better handle the issue because all of my tests on low memory/low power devices don't show any of this behavior. 

For the sake of whomever ends up handling this. please don't just throw out "Me too" or thumbs up. Give us as much info as you can on what exactly your configurations are, what hardware you are on, what versions you are running. The more you tell us, the easier it is for us to research and try to find out what is going on.

**Comment by RixzZ** [comment]
> > I can add a small video to show it...
> 
> Show your configuration please.

Here is the docker-compose.yml from the container I tested in the video. The video is just after the first 'up -d'.
```yaml
version: '3.8'

services:
  main:
    image: pihole/pihole:latest
    hostname: pi-hole.docker
    restart: always
    networks:
      external: {}
    ports:
      - 53:53/tcp
      - 53:53/udp
      - 3000:80/tcp
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - config:/etc/pihole/
      - dnsmasq:/etc/dnsmasq.d/

volumes:
  config: {}
  dnsmasq: {}

networks:
  external:
    external: true
```
I don't have NET_ADMIN capabilities since I think it's supposed to be needed for DHCP and I don't use it.

I have the WebUI mapped to port 3000 here because I usually send that through Traefik (and currently those ports are occupied by it), which I didn't use on the test to isolate the problem.

The problem occurs through Traefik, exposing the port directly, exposing the port with a different external port and even in host network mode.


> I guess I'm not understanding where there are huge slowdowns. Obviously running in a virtualized environment like docker is not going to get you bare metal performance but I'm not seeing 5 second page loads or 10x delays.

At first I thought that may be the issue, since I didn't find this problem before using it on docker, but some people reported previously that the problem doesn't occur on previous versions, so... I don't know.

**Comment by mcrapts** [comment]
> > Compare any recent version with `pihole/pihole:v5.1.2` and I am sure you will measure this 10x difference as well.
> 
> I don't.

Benchmark using Google Chrome, trying to load http://localhost/admin/queries.php.

Pihole 5.8.1:
![image](https://user-images.githubusercontent.com/16231555/116142211-eb41bc80-a6d9-11eb-9889-9c2523b11b5b.png)

Pihole 5.1.2:
![image](https://user-images.githubusercontent.com/16231555/116142338-14fae380-a6da-11eb-8f44-d2bffd35938f.png)

925 / 34 = 27x slower.

All I did was change version numbers in my `docker-compose.yml`:

```
version: "3"

# More info at https://github.com/pi-hole/docker-pi-hole/ and https://docs.pi-hole.net/
services:
  pihole:
    container_name: pihole
    image: pihole/pihole:v5.1.2
    # image: pihole/pihole:v5.8.1
    ports:
      - "53:53/tcp"
      - "53:53/udp"
      - "67:67/udp"
      - "80:80/tcp"
      - "443:443/tcp"
    environment:
      TZ: Europe/Amsterdam
      WEBPASSWORD: pihole
    # Volumes store your data between container upgrades
    volumes:
      - ./etc-pihole/:/etc/pihole/
      - ./etc-dnsmasq.d/:/etc/dnsmasq.d/
    # Recommended but not required (DHCP needs NET_ADMIN)
    #   https://github.com/pi-hole/docker-pi-hole#note-on-capabilities
    restart: unless-stopped```

Can you share the load times you are experiencing? I think it has been demonstrated numerous times a lot of people are experiencing this exact same issue, so if measure a ~30ms load time with v5.8.1 where others are experiencing 1s load times, it's definitely worth investigating what's causing the issue.

**Comment by kgillibrand** [comment]
I've reproduced this using the steps in the OP (using the stock docker-compose.yml). Page loads for query and dashboard are ~2 seconds but many pages seem to be affected.

The pihole versions I'm testing with are: Pi-hole v5.3.1 Web Interface v5.5 FTL v5.8.1

When reproducing I notice that an lsof command is taking awhile to complete and using a decent chunk of CPU (ranging from 50 - 100%).
Video of htop while loading the queries page. There is some video before and after loading to show that the host is not running much else:
https://user-images.githubusercontent.com/16925791/116804404-752fc200-aaec-11eb-9c63-bec833c38600.mp4

And a screenshot of the Chrome network page after this test was done:
![Capture2](https://user-images.githubusercontent.com/16925791/116804139-7233d200-aaea-11eb-9188-f40393c6693c.PNG)

When I run the lsof command manually inside the container it takes ~2 seconds which accounts for most of the slowness. The lsof CPU usage is roughly the same as during the page load (50-100%):
`# time lsof -Pni:53
COMMAND   PID USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
pihole-FT 400 root    4u  IPv4 412472      0t0  UDP *:53
pihole-FT 400 root    5u  IPv4 412473      0t0  TCP *:53 (LISTEN)
pihole-FT 400 root    6u  IPv6 412474      0t0  UDP *:53
pihole-FT 400 root    7u  IPv6 412475      0t0  TCP *:53 (LISTEN)

real    0m1.946s
user    0m0.775s
sys     0m1.159s
# echo $?
0`

Running the same command on the host itself is fast and there is no noticeable CPU load:
`$ time lsof -Pni:53
$ echo $?
1`

Notice that there are 4 results in the container and none in the host. The exit statuses are also different.

The host specs are:
Raspberry Pi 4B 2GB RAM
Boot drive: 400 MB/s USB3 stick
Ubuntu 21.04
Kernel: 5.11.0-1007-raspi
Docker version 20.10.2, build 20.10.2-0ubuntu2

I see other complaints online of lsof being slow: https://www.google.com/search?q=lsof+slow+linux&rlz=1C1GCEA_enCA921CA921&oq=lsof+slow+linux&aqs=chrome..69i57j69i60.1966j0j4&sourceid=chrome&ie=UTF-8
Pihole is already using the common suggestion of using the -n option to lsof though.

I don't recall this slowness happening in older versions as others have mentioned. Does anyone else notice this same behavior if they have a process monitor open while loading a page?

**Comment by nobsta** [comment]
I recognized the slow WebInterface response in last version too. But before version WebInterface v5.5 I was able to slove this by deactivating logging. After deactivating logging with "pihole logging off", WebInterface was quick again. Now with version v5.5 I have to restart my raspi. After that the WebInterface is quick again, for about one day... somewhen it starts to response slow again.

**Comment by pkoetsier** [comment]
**TL;DR:**
Try adding something like the following to docker-compose.yml:
```
    ulimits:
      nofile:
        soft: 1024
        hard: 4096
```

**The longer story**
I noticed the same thing i.e. a sluggish web UI and lsof running for more than a second when requesting a UI page. When running strace on lsof, it turned out that lsof closed all file descriptors between 3 and the open file descriptor limit (RLIMIT_NOFILE). In my pihole container the open fd limit was set to 1048576, which seems to be the default for Docker. So in my case lsof was calling close() 1048574 times. On my Debian host system, the open fd limit was 1024. When testing with the higher limit on my host system, I got the following results:

```
$ for nofile in 1024 1048576; do ulimit -Sn ${nofile}; echo -e "\nnofile=$(ulimit -Sn)"; time -p lsof -v > /dev/null 2>&1; done

nofile=1024
real 0.01
user 0.00
sys 0.01

nofile=1048576
real 1.40
user 0.79
sys 0.61

```
Quite a significant difference, but not too surprising considering that lsof called close() about 1024 times as often during the 2nd invocation.

When using "docker run" instead of docker-compose, you can  use the "--ulimit" parameter:
```
$ docker run --rm -it busybox sh -c "ulimit -Sn"
1048576
$ docker run --rm -it --ulimit nofile=1024:1024 busybox sh -c "ulimit -Sn"
1024
```

**Comment by coreyasmith** [comment]
> 
> 
> **TL;DR:**
> Try adding something like the following to docker-compose.yml:
> 
> ```
>     ulimits:
>       nofile:
>         soft: 1024
>         hard: 4096
> ```

The Web UI is very snappy after making this change. Thank you!

**Comment by kgillibrand** [comment]
I can also confirm this fixes the issue on my system using the 1024 and 4096 limits.

As a sidenote my Ubuntu 21.04 host has a nofile limit of `1024:1048576` and the container used `1048576:1048576` by default. Running the same limit as the host is also working so far.

**Comment by dschaper** [comment]
Thank you @pkoetsier! Is this something that would be good to have in the documentation/README.MD/sample compose?

**Comment by mcrapts** [comment]
Thank you a lot @pkoetsier and @coreyasmith! This fixes the issue. I can confirm the web UI is now lighting fast.  To add that, you can view the current limits easily using `ulimit`:
```
$ ulimit -Hn
1048576
$ ulimit -Sn
1024
```
I also used the hard limit of 1048576 instead of 4096 so the limits are equal to the limits on my host machine.

**Comment by pkoetsier** [comment]
> Thank you @pkoetsier! Is this something that would be good to have in the documentation/README.MD/sample compose?

I think it would be good to have these ulimits in docker-compose.yml.example.

**Comment by dschaper** [comment]
Would you be open to doing a PR with the changes? You can use the GitHub web based file editor if you're not set up to do `git` or haven't had much experience with it. We're very happy to help you with the process.

**Comment by dschaper** [comment]
And @mcrapts I humbly apologize for discounting your observations. I did that multiple times on a few different platforms. (I think you may have brought this up on Reddit in the past?). It was wrong for me to do that.

**Comment by tjengbudi** [comment]
> **TL;DR:**
> Try adding something like the following to docker-compose.yml:
> 
> ```
>     ulimits:
>       nofile:
>         soft: 1024
>         hard: 4096
> ```
> 
> **The longer story**
> I noticed the same thing i.e. a sluggish web UI and lsof running for more than a second when requesting a UI page. When running strace on lsof, it turned out that lsof closed all file descriptors between 3 and the open file descriptor limit (RLIMIT_NOFILE). In my pihole container the open fd limit was set to 1048576, which seems to be the default for Docker. So in my case lsof was calling close() 1048574 times. On my Debian host system, the open fd limit was 1024. When testing with the higher limit on my host system, I got the following results:
> 
> ```
> $ for nofile in 1024 1048576; do ulimit -Sn ${nofile}; echo -e "\nnofile=$(ulimit -Sn)"; time -p lsof -v > /dev/null 2>&1; done
> 
> nofile=1024
> real 0.01
> user 0.00
> sys 0.01
> 
> nofile=1048576
> real 1.40
> user 0.79
> sys 0.61
> ```
> 
> Quite a significant difference, but not too surprising considering that lsof called close() about 1024 times as often during the 2nd invocation.
> 
> When using "docker run" instead of docker-compose, you can use the "--ulimit" parameter:
> 
> ```
> $ docker run --rm -it busybox sh -c "ulimit -Sn"
> 1048576
> $ docker run --rm -it --ulimit nofile=1024:1024 busybox sh -c "ulimit -Sn"
> 1024
> ```

i used docker within synology system. i cannot find any location or any way to change the ulimit. have you a clue about it?

**Comment by jeff-denning** [comment]
> i used docker within synology system. i cannot find any location or any way to change the ulimit. have you a clue about it?
Should be under Environment. `nofile` is the and `1024` is the value, as an example.

**Comment by tjengbudi** [comment]
> > i used docker within synology system. i cannot find any location or any way to change the ulimit. have you a clue about it?
> 
> Should be under Environment. `nofile` is the and `1024` is the value, as an example.

trying add environment variable nofile and the value is 1024, start the docker and the ulimit is not changing

**Comment by tjengbudi** [comment]
@jeff-denning it looks like it must change at docker-composite.yml file. looks like @dschaper ask @pkoetsier to edit and create pull request. i try to edit them but cause i not familiar with docker whole system, it will be wise if someone have capability to edit and make pull request. hope it will resolve soon

**Comment by jeff-denning** [comment]
> @jeff-denning it looks like it must change at docker-composite.yml file. looks like @dschaper ask @pkoetsier to edit and create pull request. i try to edit them but cause i not familiar with docker whole system, it will be wise if someone have capability to edit and make pull request. hope it will resolve soon

The Synology Docker UI is just a frontend for docker-compose and docker-run commands. Try adding the ulimit options to environment, delete your container and start it again.

**Comment by tjengbudi** [comment]
> > @jeff-denning it looks like it must change at docker-composite.yml file. looks like @dschaper ask @pkoetsier to edit and create pull request. i try to edit them but cause i not familiar with docker whole system, it will be wise if someone have capability to edit and make pull request. hope it will resolve soon
> 
> The Synology Docker UI is just a frontend for docker-compose and docker-run commands. Try adding the ulimit options to environment, delete your container and start it again.

i might get lost with ulimit option. trying add ulimit and value is nofile=1024:1024 and it still cannot change the docker ulimit value. already delete container and start again but with no luck. i also search in google but found nothing :(

**Comment by tbraeutigam** [comment]
This might be slightly off-topic to the ulimits/docker talk. But since it pertains to the issue described (and google pointed me here), I'd like to offer another thing that can cause the same/similar WebUI slowdown:

If your `/etc/hosts` for whatever reason does not contain the hostname of the machine at `127.0.0.1`, `sudo` will be exceedingly slow (as per design it resolves the local hostname on every try). And -- I haven't investigated why -- pihole does a `sudo` on every page load.

Apologies for being a little offtopic to the issue found here, but maybe someone might appreciate this.

**Comment by wAuner** [comment]
> If your `/etc/hosts` for whatever reason does not contain the hostname of the machine at `127.0.0.1`, `sudo` will be exceedingly slow (as per design it resolves the local hostname on every try). And -- I haven't investigated why -- pihole does a `sudo` on every page load.


@tbraeutigam @jeff-denning I can confirm that this solves the issue for the docker container. I've wondered about those logs in my Synology docker setup. The webui was unusable slow. Once I've added the container name `pihole` as an additional entry for 127.0.0.1 in `/etc/hosts`it fixed the problem and everything is super fast now.

**Comment by coreyasmith** [comment]
> 
> 
> This might be slightly off-topic to the ulimits/docker talk. But since it pertains to the issue described (and google pointed me here), I'd like to offer another thing that can cause the same/similar WebUI slowdown:
> 
> If your `/etc/hosts` for whatever reason does not contain the hostname of the machine at `127.0.0.1`, `sudo` will be exceedingly slow (as per design it resolves the local hostname on every try). And -- I haven't investigated why -- pihole does a `sudo` on every page load.
> 
> Apologies for being a little offtopic to the issue found here, but maybe someone might appreciate this.

This didn't make a difference for me. My `/etc/hosts` already had the Pi's hostname, but it didn't have `pihole`. I added `127.0.0.1 pihole`, but it made no change to the responsiveness of the UI. Adding `ulimits` to `docker-compose.yml` is the only solution that's worked for me so far.

**Comment by dmce** [comment]
> ```shell
> echo "127.0.0.1 $(hostname)" >> /etc/hosts
> ```

This resolved my issue with pihole behind traefik. Each request took over 3 minutes
